### PR TITLE
Extend gpt-fast LLM dashboard to support torchao autoquant

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #140627

Summary:
We want to test autoquant on relevant LLM models

right now only llama2 and mixtral, but want to extend to more models like https://github.com/vllm-project/vllm/tree/main/vllm/model_executor/models

Test Plan:

```
                                            Llama-2-7b-chat-hf          Mixtral-8x7B-v0.1
gpt-fast int8                           112.98                              147.92
torchao autoquant                  87.41                               85.90
torchao autoquantv2             131.12                                79.59
```
 

https://hud.pytorch.org/benchmark/llms?repoName=pytorch%2Fpytorch

in pytorch/benchmarks/gpt_fast
```
python benchmark.py
```

output:
```
Loading model Llama-2-7b-chat-hf
Using int8 weight-only quantization!
Time to load model: 2.80 seconds
Compilation time: 170.24 seconds
Average tokens/sec: 112.98 tokens/sec
Average bandwidth achieved: 746.86 GB/s
Memory used: 7.95 GB

Loading model Mixtral-8x7B-v0.1
Using int8 weight-only quantization!
Time to load model: 0.24 seconds
Compilation time: 181.81 seconds
Average tokens/sec: 147.92 tokens/sec
Average bandwidth achieved: 953.06 GB/s
Memory used: 32.45 GB

Loading model Llama-2-7b-chat-hf
Time to load model: 0.11 seconds
Using autoquant
Compilation time: 109.31 seconds
Average tokens/sec: 87.17 tokens/sec
Average bandwidth achieved: 1151.86 GB/s
Memory used: 32.45 GB

Loading model Llama-2-7b-chat-hf
Time to load model: 0.11 seconds
Compilation time: 48.08 seconds
Average tokens/sec: 87.41 tokens/sec
Average bandwidth achieved: 1155.05 GB/s
Memory used: 36.86 GB

Loading model Mixtral-8x7B-v0.1
Time to load model: 0.20 seconds
Using autoquant
Compilation time: 47.32 seconds
Average tokens/sec: 85.90 tokens/sec
Average bandwidth achieved: 1106.37 GB/s
Memory used: 66.81 GB


local test (autoquant v2):
Loading model Mixtral-8x7B-v0.1
Compilation time: 124.40 seconds
Average tokens/sec: 90.41 tokens/sec
Average bandwidth achieved: 1164.47 GB/s
Memory used: 53.91 GB

Loading model Llama-2-7b-chat-hf
TODO

```

gpt_fast_benchmark.csv:
```
name,metric,target,actual,dtype,device,arch,is_model
Llama-2-7b-chat-hf,token_per_sec,144,112.98,int8,cuda,NVIDIA PG509-210,True
Llama-2-7b-chat-hf,memory_bandwidth(GB/s),957,746.86,int8,cuda,NVIDIA PG509-210,True
Llama-2-7b-chat-hf,compilation_time(s),136,170.24,int8,cuda,NVIDIA PG509-210,True
Mixtral-8x7B-v0.1,token_per_sec,175,147.92,int8,cuda,NVIDIA PG509-210,True
Mixtral-8x7B-v0.1,memory_bandwidth(GB/s),1130,953.06,int8,cuda,NVIDIA PG509-210,True
Mixtral-8x7B-v0.1,compilation_time(s),133,181.81,int8,cuda,NVIDIA PG509-210,True
gemv,memory_bandwidth(GB/s),870,867.06,int8,cuda,NVIDIA PG509-210,False
gemv,memory_bandwidth(GB/s),990,1092.43,bfloat16,cuda,NVIDIA PG509-210,False
layer_norm,memory_bandwidth(GB/s),950,573.57,bfloat16,cuda,NVIDIA PG509-210,False
Llama-2-7b-chat-hf,token_per_sec,144,87.17,autoquant,cuda,NVIDIA PG509-210,True
Llama-2-7b-chat-hf,memory_bandwidth(GB/s),957,1151.86,autoquant,cuda,NVIDIA PG509-210,True
Llama-2-7b-chat-hf,compilation_time(s),136,109.31,autoquant,cuda,NVIDIA PG509-210,True
gather_gemv,memory_bandwidth(GB/s),990,945.38,int8,cuda,NVIDIA PG509-210,False
gather_gemv,memory_bandwidth(GB/s),1060,1188.29,bfloat16,cuda,NVIDIA PG509-210,False
mlp_layer_norm_gelu,flops_utilization,0.8,0.82,bfloat16,cuda,NVIDIA PG509-210,False
Llama-2-7b-chat-hf,token_per_sec,94,87.41,bfloat16,cuda,NVIDIA PG509-210,True
Llama-2-7b-chat-hf,memory_bandwidth(GB/s),1253,1155.05,bfloat16,cuda,NVIDIA PG509-210,True
Llama-2-7b-chat-hf,compilation_time(s),133,48.08,bfloat16,cuda,NVIDIA PG509-210,True
Mixtral-8x7B-v0.1,token_per_sec,175,85.90,autoquant,cuda,NVIDIA PG509-210,True
Mixtral-8x7B-v0.1,memory_bandwidth(GB/s),1130,1106.37,autoquant,cuda,NVIDIA PG509-210,True
Mixtral-8x7B-v0.1,compilation_time(s),133,47.32,autoquant,cuda,NVIDIA PG509-210,True
```
Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

**Comment by huydhn** [comment]
It looks there there are still some failures on both A100 and x86 run:

* https://github.com/pytorch/pytorch/actions/runs/11960091588/job/33347870789
* https://github.com/pytorch/pytorch/actions/runs/11960123335/job/33346846517

**Comment by jerryzh168** [comment]
not sure about the first issue, but second one could be resolved with an update of torchao, just updated

**Comment by huydhn** [comment]
I'm also not sure what happens yet, building ao on CUDA fails, maybe we can try to reproduce this on a regular GPU CUDA runner instead of A100

**Comment by msaroufim** [comment]
The errors you're seeing are because you're building AO for very old SMs where things like bfloat16 are not supported

You can pass in optional flags to nvcc and be explicity in passing newer `gencode=arch=compute_80,code=sm_80"` so that you never build custom extensions on older hardware https://github.com/pytorch/ao/blob/main/setup.py#L82

To test locally you can probably repro most of these breakages if you try building AO from source on google colab using the default T4 machines that they provide

**Comment by huydhn** [comment]
It could be something wrong with the build script, as I see sm80 is set here https://github.com/pytorch/pytorch/blob/main/.github/workflows/inductor-micro-benchmark.yml#L50

**Comment by huydhn** [comment]
I found the issue when building ao. The problem is that it's done during testing phase in `_linux_test.yml` workflow and it doesn't set `TORCH_CUDA_ARCH_LIST` at all. The default to the value set in the Docker image https://github.com/pytorch/pytorch/blob/main/.ci/docker/ubuntu-cuda/Dockerfile#L139 is used, which fails the build.  The failure can be reproduced locally on devgpu by running `TORCH_CUDA_ARCH_LIST=Maxwell pip install "git+https://github.com/pytorch/ao.git@ca52cdc88f608e8df504e2131a97c23074c2e198"`

The fix is to set TORCH_CUDA_ARCH_LIST in the test script, and a similar fix has been done elsewhere, i.e. https://github.com/pytorch/pytorch/blob/main/.ci/pytorch/test.sh#L1462.  So, the same needs to be done here.

File: .ci/pytorch/test.sh [hunk 1]
```diff
@@ -615,6 +615,7 @@ test_single_dynamo_benchmark() {
 }
 
 test_inductor_micro_benchmark() {
+  install_torchao
```
Comment (huydhn) [inline .ci/pytorch/test.sh:618]: We need to `TORCH_CUDA_ARCH_LIST` here in the same way as https://github.com/pytorch/pytorch/blob/main/.ci/pytorch/test.sh#L1462 to successfully build ao.

```suggestion
  if [[ "$BUILD_ENVIRONMENT" == *cuda* ]]; then
    export TORCH_CUDA_ARCH_LIST="8.0;8.6"
  fi  
  install_torchao
```

**Comment by jerryzh168** [comment]
waiting for https://github.com/pytorch/pytorch/issues/141498 to be fixed, or we could ignore that and land

**Review by huydhn — APPROVED** [review]

**Pull request closed without merging**